src/librustc/hir/mod.rs

```diff
@@ -961,22 +956,24 @@ pub enum Expr_ {
     ExprRepeat(P<Expr>, P<Expr>),
 }
 
-/// The explicit Self type in a "qualified path". The actual
-/// path, including the trait and the associated item, is stored
-/// separately. `position` represents the index of the associated
-/// item qualified with this Self type.
-///
-///     <HirVec<T> as a::b::Trait>::AssociatedItem
-///      ^~~~~     ~~~~~~~~~~~~~~^
-///      ty        position = 3
-///
-///     <HirVec<T>>::AssociatedItem
-///      ^~~~~    ^
-///      ty       position = 0
+/// Optionally `Self`-qualified value/type path or associated extension.
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct QSelf {
-    pub ty: P<Ty>,
-    pub position: usize,
+pub enum QPath {
+    /// Path to a definition, optionally "fully-qualified" with a `Self` type
+    /// if it points to an associated item, e.g. `<Vec<T> as Trait>::method`.
+    Regular(Option<P<Ty>>, P<Path>),
+
+    /// Associated path extension, e.g. `<T>::default` or `<T>::Output`.
+    /// UFCS source paths can desugar into this, with `Vec::new` turning into
+    /// `<Vec>::new`, and `T::X::Y::method` into `<<<T>::X>::Y>::method`,
+    /// the inner nodes being each one `TyPath(QPath::Extend(..))`.
+    Extend(P<Ty>, P<PathSegment>)
```

Comment:
Sadly none of those suggest "associated item" (other than `TypeBased`, I suppose) - I used to have `Project` instead of `Extend` although I changed it because @nikomatsakis mentioned it ("projection" requires a trait everywhere else we use that term).

That said, I like `Resolved` instead of `Regular`. How about `Resolved` / `TypeBased`?
Or `TypeLookup` or even `TypeAssociated` although that's maybe a bit to long.